packages/smart-contracts/src/contracts/SwapToPay.sol

```diff
@@ -0,0 +1,259 @@
+ * 
+ * EXEC transferMkrFromDaiWithReference
+ * _to: 0xAa0c45D2877373ad1AB2aa5Eab15563301e9b7b3
+ * _amount: 10000000000000000000 (10 FAU)
+ * _amountInMax: 10300300000000000000 (10.3003 CTBK = 3% slippage of the total amount)
+ * _path: [0x995d6a8c21f24be1dd04e105dd0d83758343e258, 0xfab46e002bbf0b4509813474841e0716e6730136]
+ * _paymentReference: 0x202009030000
+ * _feeAmount: 10000000000000000 (0.01 FAU = 0.1%)
+ * _feeAddress: 0x61076Da38517be36d433E3fF8D6875B87880Ba56
+ * _deadline: 1600136572 (15th of Sep)
+
+/**
+ * @title SwapToPay
+ * @notice This contract swaps ERC20 tokens before paying a request thanks to a payment proxy
+  */
+contract SwapToPay {
+
+  IUniswapV2Router02 public uniswapRouter;
+  IFeePaymentNetwork public paymentProxy;
+  address public admin;
+  
+
+  constructor(address _uniswapRouterAddress, address _paymentProxyAddress) public {
+    uniswapRouter = IUniswapV2Router02(_uniswapRouterAddress);
+    paymentProxy = IFeePaymentNetwork(_paymentProxyAddress);
+    admin = msg.sender;
+  }
+
+  function approvePaymentProxyToSpend(address erc20Address) public {
+    IERC20 erc20 = IERC20(erc20Address);
+    uint256 max = 2**256 - 1;
+    erc20.approve(address(paymentProxy), max);
+  }
+  
+  function approveRouterToSpend(address erc20Address) public {
+    IERC20 erc20 = IERC20(erc20Address);
+    uint256 max = 2**256 - 1;
+    erc20.approve(address(uniswapRouter), max);
+  }
+  
+  /**
+  * @notice Performs a token swap between a payment currency and a request currency, and then
+  *         calls a payment proxy to pay the request, including fees.
+  * @param _to Transfer recipient = request issuer
+  * @param _amount Amount to transfer in request currency
+  * @param _amountInMax Maximum amount allowed to spend for currency swap, in payment currency.
+            This amount should take into account the fees.
+    @param _path, path of ERC20 tokens to swap from requestedToken to spentToken. The first 
+            address of the path should be the payment currency. The last element should be the
+            request currency.
+  * @param _paymentReference Reference of the payment related
+  * @param _feeAmount Amount of the fee in request currency
+  * @param _feeAddress Where to pay the fee
+  * @param _deadline Deadline for the swap to be valid
+  *
+  * TODO test the behaviour when the deadline is passed
+  */
+  function transferMkrFromDaiWithReference(
+    address _to,
+    uint256 _amount, // requestedToken
+    uint256 _amountInMax, // spentToken
+    address[] calldata _path, // from requestedToken to spentToken
+    bytes calldata _paymentReference,
+    uint256 _feeAmount,
+    address _feeAddress,
+    uint256 _deadline
+  )
+    external
+  {
+    IERC20 spentToken = IERC20(_path[0]);
+    IERC20 requestedToken = IERC20(_path[_path.length-1]);
+    
+    uint256 requestedTotalAmount = _amount + _feeAmount;
+
+    spentToken.transferFrom(msg.sender, address(this), _amountInMax);
```

Comment:
We probably want to use the SafeTransfer function from the fee proxy smart contract here too, to handle non-standard tokens.

Reply:
Also, we need to require.

Reply:
Thanks, done! I have moved the safeTransferFrom to a library for next contracts of this kind. I don't think it's a good idea to update the ProxyFee contract in this direction, but for a later update!